### PR TITLE
Fix most tests in `ResolveTest_1_5.java`

### PR DESCRIPTION
**Pull request by datho7561**
- Fix code select when the selection covers multiple nodes
- Fix code select for type definitions with type parameters
  (fallback to IJavaElement instead of resolving bindings)
- Fix code select for static imports of overloaded methods
- Fix code select for anonymous class contructors where the constructor
  resides in the parent type

### CODE REVIEW

**Comment by datho7561** [comment]
@mickaelistria were you trying to rebase? It seems like the commit history got a little messed up. I can push a rebase of the commit on the branch if this is the case.

**Comment by mickaelistria** [comment]
Yes, I've squashed and rebased a bunch of commits. The best approach now is most probably that you checkout the latest HEAD of dom-based-operations branch and cherry-pick the particular commit you want to add and `push --force` it.

**Comment by datho7561** [comment]
Should be down to two failing tests in the suite now.

The two remaining tests are related to a statement that the parser isn't able to recover. We might need to do some work on eclipse.jdt.core in order to fix it, such as move some code recovery hack from the completion parser down to the assist parser.

**Comment by mickaelistria** [comment]
I see it's a lot of extra code. Are parts of it some workarounds for upstream issues (eg missing binding->element resolution in JDT)?

**Comment by datho7561** [comment]
> Are parts of it some workarounds for upstream issues

Not really.

For the importing an overloaded method, the API for resolving the binding only permits returning one binding, so not much can be done without changing the API.

For the case of resolving the constructor in an anonymous class declaration, the existing API is documented as only resolving the binding for constructors declared in the anonymous class. There is no API that handles walking up through the supertypes to search for the constructor. We could look into providing such an API, but it would be a new API.

I'll read through my code here again. I've noticed that there is some code I can remove. Also, it seems I've introduced some regressions.

**Comment by datho7561** [comment]
I don't know why there are regressions. Everything I've changed should be behind the DOM_BASED_OPERATIONS flag. I'll have to look deeper into the failures Friday.

File: org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/CompilationUnit.java [hunk 1]
```diff
@@ -610,10 +650,12 @@ static IBinding resolveBinding(ASTNode node) {
 	if (node instanceof ConstructorInvocation constructor) {
 		return constructor.resolveConstructorBinding();
 	}
+	if (node instanceof org.eclipse.jdt.core.dom.Annotation annotation) {
+		return annotation.resolveTypeBinding();
+	}
```
Comment (mickaelistria) [inline org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/CompilationUnit.java]: Can you maybe just submit this in a dedicated PR. It shouldn't cause any issue and will fix a couple bugs.

Comment (datho7561) [reply org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/CompilationUnit.java]: Here it is #112

**Comment by datho7561** [comment]
Finally dug through the test results to see the difference between the HEAD of dom-based-operations and this branch, because I was expecting the #failures to go down, whereas it actually remains constant. Here are the regressions I need to address:

```
JavaSearch15Tests564049_003
ResolveTests.testClass6
ResolveTests.testInterface
ResolveTests.testInterfaceX
ResolveTests.testLocalClass6
ResolveTests.testSecondaryTypes
TypeHierarchyTests.testBugGh269
TypeResolveTests.testBug575503
```

**Comment by datho7561** [comment]
@mickaelistria if you have time, could you please take another look at this PR? It should take the test failures from 152 to 143, and it won't increase the errors.

**Comment by mickaelistria** [comment]
Cool, thanks!
